### PR TITLE
[E2E] Adds test harness

### LINKED ISSUE #580: [E2E] E2E test harness with LocalNet
## Objective

The first step to developing our E2E test suite is to get the minimal harness that wires together all the right pieces and starts the documentation for everything. This ticket is for that minimal initial harness and documentation.

Should roughly use a LocalNet cluster to spin up a cluster-manager node and call commands on that node since it is primed as a validator already. This approach should allow us the flexibility to swap out Kubernetes planes underneath the E2E tests without any changes.

## Origin Document

#466 

## Goals

- Integrate with the LocalNet in a way that can be easily extended to DevNet
- Create the harness for writing and developing with E2E tests on LocalNet
- Allow creation of pod resources from the E2E tests themselves.

## Deliverable

- [ ] Run Gherkin tests against LocalNet that can run pocket binary calls.
- [ ] Documentation outlining what the E2E tests do, how to write new ones, and how they work under the hood.

## Non-goals / Non-deliverables

- Writing a full network simulator
- Writing a comprehensive test suite
- Integrate CI/CD with the E2E tests

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **Task specific tests or benchmarks**: `make test_e2e`
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @dylanlott

### PR DESCRIPTION
**Pull request by dylanlott**
## Description

This PR adds the E2E test harness with an example of how it can drive around a KubeClient to test  commands in a Pod against the built binary. It also adds a button to the Tilt UI that manually triggers the E2E test suite to run for development purposes.

## Issue

FIxes #580 as part of a set of changes addressed by #466

## Type of change

Please mark the relevant option(s):

- [x] New fexature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Adds the e2e/tests directory.
- Adds the godog test suite runner in the e2e/tests subdirectory. 
- Adds a simple root feature that tests the help output of the CLI as an example.
- Updates the Tiltfile for LocalNet to include a button for manually triggering e2e tests.

## Testing

- [x] `make develop_test`
- [ ] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: e2e/tests/root.feature [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Feature: Root Namespace
+
+  Scenario: User Needs Help
+    Given the user has a pocket client 
+    When the user runs the command "help"
+    Then the user should be able to see standard output containing "Available Commands"
+    And the pocket client should have exited without error
```
Comment (dylanlott) [inline e2e/tests/root.feature:7]: This test format was copied form the v0 tests to show feature parity.

File: e2e/tests/kube_client.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
+//go:build e2e
```
Comment (dylanlott) [inline e2e/tests/kube_client.go:1]: Because the e2e tests depend on a Kubernetes environment to be available, we isolate the e2e tests with a build tag so they're ignored unless the test command is run with `-tags=e2e`. Issue #581 covers adding the E2E tests to our Github Actions.

File: Makefile [hunk 3]
```diff
@@ -321,6 +321,10 @@ generate_node_state_machine_diagram: ## (Re)generates the Node State Machine dia
 test_all: ## Run all go unit tests
 	go test -p 1 -count=1 ./...
 
+.PHONY: test_e2e
+test_e2e: ## Assumes that `localnet_up` is running *warm*.
```
Comment (Olshansk) [inline Makefile:325]: 1. Try running `make` from the root of the repo. The comment is actually the `help` command and not just a makefile comment.

2. **Optional NIT:** Is there a way to add a small check to validate that it's up and running? Potentially another target making a small RPC call? See `docker_check` as an example

File: build/localnet/Tiltfile [hunk 4]
```diff
@@ -199,3 +200,8 @@ k8s_resource(
     port_forwards=["42000:3000"],
     labels=["monitoring"]
 )
+
+test_go('e2e-tests', '../../e2e/tests', '.',
```
Comment (Olshansk) [inline build/localnet/Tiltfile:204]: Some of the other functions in this file accept a `root_dir` and then append the extension. Thoughts on a similar approach?

File: e2e/tests/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+e2e tests
+=========
+
+> tl; dr - `make localnet_up` and then `make test_e2e`
```
Comment (Olshansk) [inline e2e/tests/README.md:4]:
```suggestion
# E2E Tests

> tl; dr - `make localnet_up` and then `make test_e2e` 
```

File: e2e/tests/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
+e2e tests
+=========
+
+> tl; dr - `make localnet_up` and then `make test_e2e` 
+
```
Comment (Olshansk) [inline e2e/tests/README.md:5]: Consider using [1] to create a table of contents.

[1] https://marketplace.visualstudio.com/items?itemName=huntertran.auto-markdown-toc

File: e2e/tests/README.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+You can also click the `e2e-tests` button in the Tilt UI, which is handy during development.
```
Comment (Olshansk) [inline e2e/tests/README.md:6]: Insert a screenshot?

File: e2e/tests/README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+## Design & Architecture
+
+The test suite is located in `e2e/tests` and it contains a set of Cucumber Gherkin feature files and the associated Go tests to run them. This is a test file that lays out the BDD pattern using the client's help function as an example.
+
+### Implementation 
+
+`make test_e2e` sees any files named with the pattern `*.feature` in `e2e/tests` and runs them with [godog](https://github.com/cucumber/godog), the Go test runner for Cucumber tests. We indirectly depend on the LocalNet to be up and running for our test suite to use. We call commands to steer around a container by calling `kubectl exec` and targeting a pod in the cluster by name. We are able to pass commands to that client and then assert on the behavior of that program.
```
Comment (Olshansk) [inline e2e/tests/README.md:14]:
```suggestion
`make test_e2e` sees any files named with the pattern `*.feature` in `e2e/tests` and runs them with [godog](https://github.com/cucumber/godog), the Go test runner for Cucumber tests. There is an indirect dependance on the LocalNet to be up and running for our test suite to use. We call commands to steer around a container by calling `kubectl exec` and targeting a pod in the cluster by name. We are able to pass commands to that client and then assert on the behavior of that program.
```

Try to avoid using `we` and make it more 3rd person.

File: e2e/tests/README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    ├── README.md
+    ├── kube_client.go
+    ├── root.feature
+    └── steps_init_test.go
+
+2 directories, 4 files
+```
+
+`kube_client.go` defines the KubeClient that grabs a Pod and drives it around for the tests.
+`steps_init_test.go` registers the handler functions and runs the test suite.
```
Comment (Olshansk) [inline e2e/tests/README.md:28]: This will highlight things appropriately.
```suggestion
```bash
e2e/
└── tests
    ├── README.md
    ├── kube_client.go # defines the KubeClient that grabs a Pod and drives it around for the tests.
    ├── root.feature # registers the handler functions and runs the test suite.
    └── steps_init_test.go

2 directories, 4 files
```

File: e2e/tests/README.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+### KubeClient and Kubernetes Clientsets
+
+The KubeClient wraps the command and saves a reference to the last run command's result. The local environment's Kube client configuration is used to call commands. The Kubernetes Clientset can be changed in the test suite by swapping out the client that is defined at the package level. Multiple configurations can be maintained to test multiple environments and they can be swapped out arbitrarily.
```
Comment (Olshansk) [inline e2e/tests/README.md:32]: @okdas Can I get your feedback here. I'm very weary of enforcing my own style, but I feel like this could be easier to read as 4 separate bullet points instead of a paragraph.

Looking for an extra opinion (outside my own and Dylan's).

File: e2e/tests/README.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+### Separation of Concerns
+
+The hardest architectural problem is correctly waiting for services so that we can call the commands at the correct moment, but doing so in the package without altering the LocalNet component. We should indirectly depend on the LocalNet but not modify it, since we consider it a separate concern. To achieve this, we will need to query the Kubernetes Clientset that we acquire from the local configuration and repeatedly poll that for changes until all services our tests rely on are running.
```
Comment (Olshansk) [inline e2e/tests/README.md:36]: this is good to know, but I feel like there is missing context.

I'm trying to put myself in the shoes of someone other than me or you (who has not been thinking about this problem). Is this TMI? What were your thoughts of the v0 E2E doc?

File: e2e/tests/kube_client.go (hunk 2 above)
Comment (Olshansk) [reply e2e/tests/kube_client.go:1]: Thoughts on copy-pasting your comment in the README?

File: e2e/tests/kube_client.go [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+package e2e
+
+import (
+	"os/exec"
+)
+
+type CommandResult struct {
```
Comment (Olshansk) [inline e2e/tests/kube_client.go:9]: Does this struct need to be exposed outside of this package?

1. We can potentially downscope until it's necessary? Ditto for the internal types.

File: e2e/tests/kube_client.go [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+var _ PocketClient = &KubeClient{}
+
+// PocketClient is a single function interface for interacting with a node.
+type PocketClient interface {
+	RunCommand(...string) (*CommandResult, error)
+}
+
+// KubeClient saves a reference to a command
+type KubeClient struct {
```
Comment (Olshansk) [inline e2e/tests/kube_client.go:23]: Ditto, whenever I see things publically exposed (capitalized), I put myself in the shoes of a new developer in the codebase and try to understand: "how should I be using this?"

Downscope until necessary. Given that it implements the `PocketClient` interface, I don't think that'll be needed.

File: e2e/tests/kube_client.go [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+	result *CommandResult // stores the result of the last command that was run.
+}
+
+// RunCommand runs a command on a KubeClient.
+func (k *KubeClient) RunCommand(args ...string) (*CommandResult, error) {
+	base := []string{"exec", "-it", "deploy/pocket-v1-cli-client", "--container", "pocket", "--", "client"}
+	args = append(base, args...)
+	cmd := exec.Command("kubectl", args...)
```
Comment (Olshansk) [inline e2e/tests/kube_client.go:31]: Why are we ignoring the error?

File: e2e/tests/root.feature (hunk 1 above)
Comment (Olshansk) [reply e2e/tests/root.feature:7]: Great starting point, especially because `help` is appropriate everywhere.

I'd suggest not going for feature parity with v0 for the real business logic. We can make it better, simpler and more effective :)

File: e2e/tests/steps_init_test.go [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+func thePocketClientShouldHaveExitedWithoutError() error {
+	return client.result.Err
+}
+
+func theUserHasAPocketClient() error {
+	_, err := client.RunCommand("help")
+	return err
+}
+
+func theUserRunsTheCommand(arg1 string) error {
```
Comment (Olshansk) [inline e2e/tests/steps_init_test.go:26]: should we just make this `args ...string` straight of the bat?

File: e2e/tests/steps_init_test.go [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return err
+	}
+	if result.Err != nil {
+		return result.Err
+	}
+	return nil
+}
+
+func theUserShouldBeAbleToSeeStandardOutputContaining(arg1 string) error {
```
Comment (Olshansk) [inline e2e/tests/steps_init_test.go:37]: Rename `arg1` to `expectedOutput` or something similar

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: build/localnet/Tiltfile (hunk 4 above)
Comment (Olshansk) [reply build/localnet/Tiltfile:204]: @okdas for a second pair of 👀

File: e2e/tests/kube_client.go (hunk 14 above)
Comment (dylanlott) [reply e2e/tests/kube_client.go:31]: Which error do you mean here?

Comment (Olshansk) [reply e2e/tests/kube_client.go:31]: The way I've seen this used in the past is:

```go
cmd := exec.Command("prog")
if errors.Is(cmd.Err, exec.ErrDot) {
	cmd.Err = nil
}
if err := cmd.Run(); err != nil {
	log.Fatal(err)
}
```

Source: https://pkg.go.dev/os/exec

Is that something we want to do or is `cmd.Output` capturing it?

**Comment by Olshansk** [comment]
@dylanlott I know you've added a lot of new code, so please resolve the comments that you've tended to and re-request a review once ready. Maybe @okdas should review as well?

**Comment by dylanlott** [comment]
Closing, addressed by #644.
